### PR TITLE
Enable Style/OpenStructUse cop

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
I would like us to enable the `Style/OpenStructUse` cop. The official RuboCop [docs](https://docs.rubocop.org/rubocop/cops_style.html) cite "performance, version compatibility, and potential security issues", but an add'l issue in this repo is that it [breaks](https://sorbet.run/#%23%20typed%3A%20true%0A%0Afoo%20%3D%20OpenStruct.new%0Afoo%5B%3Aa%5D%20%3D%20'a'%0Afoo.a) typechecking.

### CODE REVIEW

File: Library/.rubocop.yml [hunk 1]
```diff
@@ -405,9 +405,16 @@ Style/NumericLiterals:
   Exclude:
     - "**/Brewfile"
 
-# OpenStruct is a nice helper.
+# TODO: These are pre-existing violations and should be fixed.
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml:408]: How about adding `rubocop:todo` instead so more violations aren't added to these files?

File: Library/.rubocop.yml [hunk 2]
```diff
@@ -405,9 +405,16 @@ Style/NumericLiterals:
   Exclude:
     - "**/Brewfile"
 
-# OpenStruct is a nice helper.
```
Comment (MikeMcQuaid) [inline Library/.rubocop.yml:408]: I still agree with this statement! What would you think about us talking over a refactoring some of the gnarlier uses here e.g. `cli/args` before we make a decision here?

**Comment by apainintheneck** [comment]
I looked a little bit into this a while ago and using it in `cli/args` didn't seem to be a significant performance issue. Still might be nice to remove it at some point but it'd be a major refactor.

File: Library/.rubocop.yml (hunk 2 above)
Comment (apainintheneck) [reply Library/.rubocop.yml:408]: That or the `Tab` class which also inherits from `OpenStruct`.

Comment (MikeMcQuaid) [reply Library/.rubocop.yml:408]: Yes, good call @apainintheneck.

**Comment by Bo98** [comment]
I wouldn't be worried too much about CLI args too much as that's always going to be an auto-gen job when it comes to typing.

**Comment by dduugg** [comment]
@apainintheneck @MikeMcQuaid PTAL: I refactored away all but three uses, one in a rescue statement that I couldn't figure out how to hit, and two others that inherit from `OpenStruct` cops. (I didn't inline the `rubocop:disable` bc the accompanying comments show up in the public documentation). I don't have a plan to migrate cli/args or tab in the foreseeable future, but still feel this change is a net improvement. LMK what you think.

File: Library/Homebrew/completions.rb [hunk 3]
```diff
@@ -12,6 +12,14 @@ module Homebrew
   module Completions
     extend T::Sig
 
+    Variables = Struct.new(
+      :aliases,
+      :builtin_command_descriptions,
+      :completion_functions,
+      :function_mappings,
+      keyword_init: true,
```
Comment (apainintheneck) [inline Library/Homebrew/completions.rb:20]: I just learned about `keyword_init: true` two days ago and it's the best thing ever.

**Review by apainintheneck — APPROVED** [review]
LGTM, it's definitely a net positive like you said.

File: Library/Homebrew/completions.rb [hunk 4]
```diff
@@ -12,6 +12,14 @@ module Homebrew
   module Completions
     extend T::Sig
 
+    Variables = Struct.new(
```
Comment (MikeMcQuaid) [inline Library/Homebrew/completions.rb:15]: Stupid question perhaps: what makes a `Struct` ok but an `OpenStruct` not?

File: Library/Homebrew/cmd/uses.rb [hunk 5]
```diff
@@ -67,7 +67,10 @@ def uses
       opoo e
       used_formulae_missing = true
       # If the formula doesn't exist: fake the needed formula object name.
+      # This is a legacy use of OpenStruct that should be refactored.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/uses.rb:70]: Ideally: maybe suggest what would be a better fit than `OpenStruct` here (and elsewhere)?

File: Library/Homebrew/manpages.rb [hunk 6]
```diff
@@ -14,6 +14,21 @@ module Homebrew
   module Manpages
     extend T::Sig
 
+    Variables = Struct.new(
```
Comment (MikeMcQuaid) [inline Library/Homebrew/manpages.rb:17]: Could/should this just be a `Hash` instead?

File: Library/Homebrew/completions.rb (hunk 4 above)
Comment (dduugg) [reply Library/Homebrew/completions.rb:15]: An `OpenStruct` accepts any method, returning `nil` by default (i think of it as a Hash that converts dot access to bracket access, not sure how accurate that is under the hood). A `Struct` only defines methods for the args in its constructor:

https://sorbet.run/#%23%20typed%3A%20true%0A%0AFoo%20%3D%20Struct.new%28%3Afoo%29%0A%0Abar%20%3D%20Foo.new%28'a'%29%0A%0Abar.foo%0Abar.baz%0A%0Aos%20%3D%20OpenStruct.new%0Aos.foo%20%23%20valid%2C%20returns%20nil%0A

File: Library/Homebrew/manpages.rb (hunk 6 above)
Comment (dduugg) [reply Library/Homebrew/manpages.rb:17]: I don't think so, though I don't have a lot of experience with ERB generation. This gives us the benefit of erroring if any of the properties are fat-fingered below.

File: Library/Homebrew/completions.rb (hunk 4 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/completions.rb:15]: Ok, gotcha, thanks!

File: Library/Homebrew/manpages.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/manpages.rb:17]: Ok, thanks!

**Pull request merged**